File: source/includes/fts/embedded-document/nested-array-query.js
const { MongoClient } = require("mongodb");

// connect to your Atlas cluster
const uri = "<connection-string>";
    
const client = new MongoClient(uri);

async function run() {
    try {
        await client.connect();

        // set namespace
        const database = client.db("local_school_district");
        const coll = database.collection("schools");

        // define pipeline
        const agg = [
          {
            '$search': {
              'index': 'embedded-documents-tutorial', 
              'embeddedDocument': {
                'path': 'teachers', 
                'operator': {
                  'compound': {
                    'must': [
                      {
                        'embeddedDocument': {
                          'path': 'teachers.classes', 
                          'operator': {
                            'compound': {
                              'must': [
                                {
                                  'text': {
                                    'path': 'teachers.classes.grade', 
                                    'query': '12th'
                                  }
                                }, {
                                  'text': {
                                    'path': 'teachers.classes.subject', 
                                    'query': 'science'
                                  }
                                }
                              ]
                            }
                          }
                        }
                      }
                    ], 
                    'should': [
                      {
                        'text': {
                          'path': 'teachers.last', 
                          'query': 'smith'
                        }
                      }
                    ]
                  }
                }
              }, 
              'highlight': {
                'path': 'teachers.classes.subject'
              }
            }
          }, {
            '$project': {
              '_id': 1, 
              'teachers': 1, 
              'score': {
                '$meta': 'searchScore'
              }, 
              'highlights': {
                '$meta': 'searchHighlights'
              }
            }
          }
        ];
        
        // run pipeline
        const result = coll.aggregate(agg);

        // print results
        await result.forEach((doc) => console.dir(JSON.stringify(doc)));
    } finally {
        await client.close();
    }
}
run().catch(console.dir);

File: source/includes/fts/embedded-document/nested-array-query.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Bson.Serialization.Conventions;
using MongoDB.Driver;
using MongoDB.Driver.Search;

public class NestedArrayExample
{
    private const string MongoConnectionString = "<connection-string>";

    public static void Main(string[] args)
    {
        // allow automapping of the camelCase database fields to our SchoolDocument
        var camelCaseConvention = new ConventionPack { new CamelCaseElementNameConvention() };
        ConventionRegistry.Register("CamelCase", camelCaseConvention, type => true);

        // connect to your Atlas cluster
        var mongoClient = new MongoClient(MongoConnectionString);
        var districtSchoolsDatabase = mongoClient.GetDatabase("local_school_district");
        var schoolsCollection = districtSchoolsDatabase.GetCollection<SchoolDocument>("schools");

        // define variables for query 
        var compoundQuery = Builders<TeacherDocument>.Search.Compound()
            .Must(Builders<TeacherDocument>.Search.Text(teacher => teacher.First, "John"))
            .Should(Builders<TeacherDocument>.Search.Text(teacher => teacher.Last, "Smith"));
        var opts = new SearchHighlightOptions<SchoolDocument>(school => school.Teachers.Select(teacher => teacher.Last));;

        // define and run pipeline
        var results = schoolsCollection.Aggregate()
            .Search(Builders<SchoolDocument>.Search.EmbeddedDocument(
                school => school.Teachers, compoundQuery), opts,
                indexName: "embedded-documents-tutorial"
            )
            .Project<SchoolDocument>(Builders<SchoolDocument>.Projection
                .Include(school => school.Name)
                .Include(school => school.Mascot)
                .Include(school => school.Teachers)
                .MetaSearchScore(school => school.Score)
                .MetaSearchHighlights("highlights"))
            .ToList();

        // print results
        foreach (var school in results)
        {
            Console.WriteLine(school.ToJson());
        }
    }
}

[BsonIgnoreExtraElements]
public class SchoolDocument
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Mascot { get; set; }
    public TeacherDocument[] Teachers { get; set; }
    [BsonElement("highlights")]
    public List<SearchHighlight> Highlights { get; set; }
    public double Score { get; set; }
}

[BsonIgnoreExtraElements]
public class TeacherDocument
{
    public string First { get; set; }
    public string Last { get; set; }
    public ClassDocument[] Classes { get; set; }
}

[BsonIgnoreExtraElements]
public class ClassDocument
{
    public string Subject { get; set; }
    public string Grade { get; set; }
}